// @varlet/ui@2.2.1/lib/utils/shared.js
"use strict";

exports.__esModule = true;
exports.padStart = exports.linear = exports.isHTMLSupportVideo = exports.isHTMLSupportImage = exports.getGlobalThis = exports.easeInOutCubic = exports.dt = exports.cubic = exports.createCache = void 0;

var _shared = require("@varlet/shared");

var isHTMLSupportImage = val => {
  if (val == null) {
    return false;
  }

  return val.startsWith('data:image') || /\.(png|jpg|gif|jpeg|svg|webp)$/.test(val);
};

exports.isHTMLSupportImage = isHTMLSupportImage;

var isHTMLSupportVideo = val => {
  if (val == null) {
    return false;
  }

  return val.startsWith('data:video') || /\.(mp4|webm|ogg)$/.test(val);
};

exports.isHTMLSupportVideo = isHTMLSupportVideo;

var createCache = max => {
  var cache = [];
  return {
    cache,

    has(key) {
      return this.cache.includes(key);
    },

    add(key) {
      if (this.has(key)) {
        return;
      }

      this.cache.length === max && cache.shift();
      this.cache.push(key);
    },

    remove(key) {
      this.has(key) && (0, _shared.removeItem)(this.cache, key);
    },

    clear() {
      this.cache.length = 0;
    }

  };
};

exports.createCache = createCache;

var linear = value => value;

exports.linear = linear;

var cubic = value => Math.pow(value, 3);

exports.cubic = cubic;

var easeInOutCubic = value => value < 0.5 ? cubic(value * 2) / 2 : 1 - cubic((1 - value) * 2) / 2;

exports.easeInOutCubic = easeInOutCubic;

var dt = (value, defaultText) => value == null ? defaultText : value;

exports.dt = dt;

var getGlobalThis = () => {
  if (typeof globalThis !== 'undefined') return globalThis;
  if (typeof window !== 'undefined') return window;
  return typeof global !== 'undefined' ? global : self;
};

exports.getGlobalThis = getGlobalThis;

var padStart = function (str, maxLength, fillString) {
  if (str === void 0) {
    str = '';
  }

  if (fillString === void 0) {
    fillString = '';
  }

  if (str.length >= maxLength) return str;
  var len = maxLength - str.length;
  var repeatCount = Math.floor(len / fillString.length);
  return fillString.repeat(repeatCount) + fillString.slice(0, len % fillString.length) + str;
};

exports.padStart = padStart;
