// @varlet/ui/lib/snackbar/props.js
"use strict";

exports.__esModule = true;
exports.positionValidator = positionValidator;
exports.props = void 0;
exports.typeValidator = typeValidator;

var _components = require("../utils/components");

var _props = require("../loading/props");

var _index = require("./index");

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function positionValidator(position) {
  var validPositions = ['top', 'center', 'bottom'];
  return validPositions.includes(position);
}

function typeValidator(type) {
  return _index.SNACKBAR_TYPE.includes(type);
}

var props = {
  type: {
    type: String,
    validator: typeValidator
  },
  // snackbar显示的位置
  position: {
    type: String,
    default: 'top',
    validator: positionValidator
  },
  // content内容
  content: {
    type: String
  },
  // 为snackbar content添加自定义类名
  contentClass: {
    type: String
  },
  // snackbar 持续时间
  duration: {
    type: Number,
    default: 3000
  },
  // 是否将消息条内容堆叠在操作（按钮）之上
  vertical: {
    type: Boolean,
    default: false
  },
  loadingType: (0, _components.pickProps)(_props.props, 'type'),
  loadingSize: (0, _components.pickProps)(_props.props, 'size'),
  loadingRadius: (0, _components.pickProps)(_props.props, 'radius'),
  loadingColor: _extends({}, (0, _components.pickProps)(_props.props, 'color'), {
    default: 'currentColor'
  }),
  // 是否禁止滚动穿透
  lockScroll: {
    type: Boolean,
    default: false
  },
  // 控制组件可见还是隐藏
  show: {
    type: Boolean,
    default: false
  },
  // teleport
  teleport: {
    type: String,
    default: 'body'
  },
  // 是否禁止点击背景
  forbidClick: {
    type: Boolean,
    default: false
  },
  // 打开时的回调函数
  onOpen: {
    type: Function,
    default: () => {}
  },
  // 打开动画结束时的回调
  onOpened: {
    type: Function,
    default: () => {}
  },
  // 关闭时的回调函数
  onClose: {
    type: Function,
    default: () => {}
  },
  // 关闭动画结束时的回调
  onClosed: {
    type: Function,
    default: () => {}
  },
  'onUpdate:show': {
    type: Function
  },
  _update: {
    type: String
  }
};
exports.props = props;
